samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java

```diff
@@ -0,0 +1,212 @@
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.samza.context.Context;
+import org.apache.samza.operators.functions.ClosableFunction;
+import org.apache.samza.operators.functions.InitableFunction;
+import org.apache.samza.serializers.Serde;
+
+
+public abstract class BaseCouchbaseTableFunction<V> implements InitableFunction, ClosableFunction, Serializable {
+
+  // Clients
+  protected transient CouchbaseEnvironment env;
+  protected transient Cluster cluster;
+  protected transient Bucket bucket;
+
+  // Function Settings
+  protected Class<V> valueClass = null;
+  protected Serde<V> valueSerde = null;
+  protected long timeout = 0L;
+  protected TimeUnit timeUnit = null;
+  protected int ttl = 0; // default value 0 means no ttl, data will be stored forever
+
+  // Cluster Settings
+  protected List<String> clusterNodes = null;
+  protected String username = null;
+  protected String password = null;
+  protected String bucketName = null;
+
+  // Environment Settings
+  protected boolean sslEnabled = false;
+  protected boolean certAuthEnabled = false;
+  protected String sslKeystoreFile = null;
+  protected String sslKeystorePassword = null;
+  protected String sslTruststoreFile = null;
+  protected String sslTruststorePassword = null;
+  protected int bootstrapCarrierDirectPort = -1;
+  protected int bootstrapCarrierSslPort = -1;
+  protected int bootstrapHttpDirectPort = -1;
+  protected int bootstrapHttpSslPort = -1;
+
+  public BaseCouchbaseTableFunction() {
+  }
+
+  //TODO maybe we can create a builder class to create both read and write functions for the same bucket so that users don't need to type in the same things twice
+  public BaseCouchbaseTableFunction(Class<V> valueClass) {
+    this.valueClass = valueClass;
+  }
+
+  @Override
+  public void init(Context context) {
+    //TODO validation
+    DefaultCouchbaseEnvironment.Builder envBuilder = new DefaultCouchbaseEnvironment.Builder();
+    // ssl settings
+    envBuilder.sslEnabled(sslEnabled).certAuthEnabled(certAuthEnabled);
+    if (sslKeystoreFile != null) {
+      envBuilder.sslKeystoreFile(sslKeystoreFile);
+    }
+    if (sslKeystorePassword != null) {
+      envBuilder.sslKeystorePassword(sslKeystorePassword);
+    }
+    if (sslTruststoreFile != null) {
+      envBuilder.sslTruststoreFile(sslTruststoreFile);
+    }
+    if (sslTruststorePassword != null) {
+      envBuilder.sslTruststorePassword(sslTruststorePassword);
+    }
+    if (bootstrapCarrierDirectPort != -1) {
+      envBuilder.bootstrapCarrierDirectPort(bootstrapCarrierDirectPort);
+    }
+    if (bootstrapCarrierSslPort != -1) {
+      envBuilder.bootstrapCarrierSslPort(bootstrapCarrierSslPort);
+    }
+    if (bootstrapHttpDirectPort != -1) {
+      envBuilder.bootstrapHttpDirectPort(bootstrapHttpDirectPort);
+    }
+    if (bootstrapHttpSslPort != -1) {
+      envBuilder.bootstrapHttpSslPort(bootstrapHttpSslPort);
+    }
+    env = envBuilder.build();
```

Comment:
We decided cluster and env are shared across tables, how do we implement that? Perhaps bucket can be shared as well?

Reply:
Yes... Working on it. I wish bucket can also be shared.